### PR TITLE
new(plugin): Azure Redis Cache with 10 modes + discovery

### PR DESCRIPTION
**Pull request by matoy**
[root@svzam-cenpol001 mode]# /usr/lib/centreon/plugins//centreon-plugins/centreon_plugins.pl --plugin=cloud::azure::database::redis::plugin --custommode='api' --list-mode

Plugin Description:
    Check Microsoft Azure Redis Cache service.

Global Options:
    --mode  Choose a mode.

    --dyn-mode
            Specify a mode with the path (separated by '::').

    --list-mode
            List available modes.

    --mode-version
            Check minimal version of mode. If not, unknown error.

    --version
            Display plugin version.

    --custommode
            Choose a custom mode.

    --list-custommode
            List available custom modes.

    --multiple
            Multiple custom mode objects (required by some specific modes)

    --pass-manager
            Use a password manager.

Modes Meta:
   multi

Modes Available:
   cache-hits
   cache-latency
   cache-miss-rate
   cache-throughput
   connected-clients
   cpu
   discovery
   errors
   memory
   operations
   server-load

[xxx redis]# /usr/lib/centreon/plugins//centreon-plugins/centreon_plugins.pl --plugin=cloud::azure::database::redis::plugin --custommode='api' --mode='discovery' --subscription='xxx' --tenant='xxx' --client-id='xxx' --client-secret='***' --resource-group='xxx'
{"end_time":1621416611,"start_time":1621416611,"duration":0,"discovered_items":1,"results":[{"location":"westeurope","name":"xxx","resourceGroup":"xxx","type":"Microsoft.Cache/Redis","id":"/subscriptions/xxx/resourceGroups/xxx/providers/Microsoft.Cache/Redis/xxx","tags":[]}]}

[xxx mode]# /usr/lib/centreon/plugins//centreon-plugins/centreon_plugins.pl --plugin=cloud::azure::database::redis::plugin --custommode='api' --mode='cache-hits' --subscription='xxx' --tenant='xxx' --client-id='xxx' --client-secret='***' --resource-group='xxx' --resource='xxx' --warning-cache-misses='800' --critical-cache-misses='900'
OK: Instance 'xxx' Statistic 'total' Metrics Cache Misses: 426.00, Cache Hits: 494.00 | 'xxx~total#redis.cache.misses.count'=426.00;0:800;0:900;0; 'xxx~total#redis.cache.hits.count'=494.00;;;0;

[xxx mode]# /usr/lib/centreon/plugins//centreon-plugins/centreon_plugins.pl --plugin=cloud::azure::database::redis::plugin --custommode='api' --mode='cache-latency' --subscription='xxx' --tenant='xxx' --client-id='xxx' --client-secret='***' --resource-group='xxx' --resource='cdc-abcdior-redis-eu-dev' --warning-cache-latency='8000' --critical-cache-latency='9000'
OK: Instance 'cdc-abcdior-redis-eu-dev' Statistic 'average' Metrics Cache Latency: 2190.66μs | 'cdc-abcdior-redis-eu-dev~average#redis.cache.latency.microseconds'=2190.66μs;0:8000;0:9000;0;

xxx mode]# /usr/lib/centreon/plugins//centreon-plugins/centreon_plugins.pl --plugin=cloud::azure::database::redis::plugin --custommode='api' --mode='cache-miss-rate' --subscription='xxx' --tenant='xxx' --client-id='xxx' --client-secret='***' --resource-group='xxx' --resource='xxx' --warning-cache-miss-rate='40' --critical-cache-miss-rate='50'
OK: Instance 'xxx' Statistic 'average' Metrics Cache Miss Rate: 14.00% | 'xxx~average#redis.cache.miss.rate'=14.00%;0:40;0:50;0;100

[xxx mode]# /usr/lib/centreon/plugins//centreon-plugins/centreon_plugins.pl --plugin=cloud::azure::database::redis::plugin --custommode='api' --mode='connected-clients' --subscription='xxx' --tenant='xxx' --client-id='xxx' --client-secret='***' --resource-group='xxx' --resource='xxx' --warning-connected-clients='800' --critical-connected-clients='900'
OK: Instance 'xxx' Statistic 'average' Metrics Connected Clients: 25.90 | 'xxx~average#redis.cache.connected.clients.count'=25.90;0:800;0:900;0;

[xxx mode]# /usr/lib/centreon/plugins//centreon-plugins/centreon_plugins.pl --plugin=cloud::azure::database::redis::plugin --custommode='api' --mode='server-load' --subscription='xxx' --tenant='xxx' --client-id='xxx' --client-secret='***' --resource-group='xxx' --resource='xxx' --warning-server-load-percentage='80' --critical-server-load-percentage='90'
OK: Instance 'xxx' Statistic 'maximum' Metrics Server Load: 1.00% | 'xxx~maximum#redis.cache.server-load'=1.00%;0:80;0:90;0;100

[xxx mode]# /usr/lib/centreon/plugins//centreon-plugins/centreon_plugins.pl --plugin=cloud::azure::database::redis::plugin --custommode='api' --mode='operations' --subscription='xxx' --tenant='xxx' --client-id='xxx' --client-secret='***' --resource-group='xxx' --resource='xxx' --warning-operations-per-second='80' --critical-operations-per-second='90'
OK: Instance 'xxx' Statistic 'maximum' Metrics Operations per seconds: 8.80op/s | 'xxx~average#redis.cache.operations.persecond'=8.80op/s;0:80;0:90;0;

[xxx mode]# /usr/lib/centreon/plugins//centreon-plugins/centreon_plugins.pl --plugin=cloud::azure::database::redis::plugin --custommode='api' --mode='cpu' --subscription='xxx' --tenant='xxx' --client-id='xxx' --client-secret='***' --resource-group='xxx' --resource='xxx' --warning-cpu-usage-percentage='80' --critical-cpu-usage-percentage='90'
OK: Instance 'xxx' Statistic 'maximum' Metrics CPU Usage Percentage: 13.00% | 'xxx~maximum#redis.cache.cpu.usage.percentage'=13.00%;0:80;0:90;0;100

[xxx mode]# /usr/lib/centreon/plugins//centreon-plugins/centreon_plugins.pl --plugin=cloud::azure::database::redis::plugin --custommode='api' --mode='memory' --subscription='xxx' --tenant='xxx' --client-id='xxx' --client-secret='***' --resource-group='xxx' --resource='xxx' --warning-memory-usage-percentage='80' --critical-memory-usage-percentage='90'
OK: Instance 'xxx' Statistic 'maximum' Metrics Memory Usage Percentage: 1.00% | 'xxx~maximum#redis.cache.memory.usage.percentage'=1.00%;0:80;0:90;0;100

[xxx mode]# /usr/lib/centreon/plugins//centreon-plugins/centreon_plugins.pl --plugin=cloud::azure::database::redis::plugin --custommode='api' --mode='errors' --subscription='xxx' --tenant='xxx' --client-id='xxx' --client-secret='***' --resource-group='xxx' --resource='xxx' --critical-errors-count='1'
OK: Instance 'xxx' Statistic 'total' Metrics Errors: 0.00 | 'xxx~total#redis.cache.errors.count'=0.00;;;0;

[xxx mode]# /usr/lib/centreon/plugins//centreon-plugins/centreon_plugins.pl --plugin=cloud::azure::database::redis::plugin --custommode='api' --mode='cache-throughput' --subscription='xxx' --tenant='xxx' --client-id='xxx' --client-secret='***' --resource-group='xxx' --resource='xxx' --warning-cache-read-throughput='800000' --critical-cache-read-throughput='900000'
OK: Instance 'xxx' Statistic 'maximum' Metrics Cache Read Throughput: 352527.00B/s, Cache Write Throughput: 133482.00B/s | 'xxx~maximum#redis.cache.read.throughput'=352527.00B/s;0:800000;0:900000;0; 'xxx~maximum#redis.cache.write.throughput'=133482.00B/s;;;0;

### CODE REVIEW

**Comment by thibaults-centreon** [comment]
Hello @matoy thank you for this great PR!
could you please merge the 3 following modes: 
cache-hits
cache-latency
cache-miss-rate
in one ('cache-usage' for example) ? From what I see these metrics share the same calculation method so they can be fetched at once.

Afterwards, I'll probably take a deeper look in order to also include the 'redisEntreprise' types in the same Plugin (https://docs.microsoft.com/en-us/azure/azure-monitor/essentials/metrics-supported#microsoftcacheredisenterprise).

**Comment by matoy** [comment]
i don't think graphs would be easily readable because numbers are in very different scales:
hits are "count" / aggregated: total
latency is in "μs" / aggregated: avg
miss-rate is a "%" /aggregated: avg

**Comment by thibaults-centreon** [comment]
Ok, I would propose this:
- rename "cachehits" to cacheusage
- discard the miss rate mode for now as the MS documentation about the stats for this metric is wrong (in https://docs.microsoft.com/en-us/azure/azure-monitor/essentials/metrics-supported#microsoftcacheredis)
- rename "connectedtclients" to "clients"
- rename "serverload" to "load"

Aditionnally, some labels need to be adjusted, see comments in code

Thanks!

File: cloud/azure/database/redis/mode/serverload.pm [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+=item B<--warning-*>
+
+Warning threshold where '*' can be:
+'server-load-percentage'.
+
+=item B<--critical-*>
+
+Critical threshold where '*' can be:
+'server-load-percentage'.
```
Comment (thibaults-centreon) [inline cloud/azure/database/redis/mode/serverload.pm:139]:
```suggestion
=item B<--warning-server-load-percentage>

Warning threshold.

=item B<--critical-server-load-percentage>

Critical threshold.
```

File: cloud/azure/database/redis/plugin.pm [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        'cache-hits'   => 'cloud::azure::database::redis::mode::cachehits',
+        'cache-latency'   => 'cloud::azure::database::redis::mode::cachelatency',
+        'cache-miss-rate'   => 'cloud::azure::database::redis::mode::cachemissrate',
+        'connected-clients'   => 'cloud::azure::database::redis::mode::connectedclients',
+        'server-load'   => 'cloud::azure::database::redis::mode::serverload',
+        'operations'   => 'cloud::azure::database::redis::mode::operations',
+        'cpu'   => 'cloud::azure::database::redis::mode::cpu',
+        'memory'   => 'cloud::azure::database::redis::mode::memory',
+        'errors'   => 'cloud::azure::database::redis::mode::errors',
+        'cache-throughput'   => 'cloud::azure::database::redis::mode::cachethroughput'
```
Comment (thibaults-centreon) [inline cloud/azure/database/redis/plugin.pm:44]: Sort & align please
Also please rename the modes as discussed in the main conversation ;)
Thanks!

File: cloud/azure/database/redis/mode/cachehits.pm [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    }
+}
+
+1;
+
+__END__
+
+=head1 MODE
+
+Check Azure Redis Cache hits/misses.
```
Comment (thibaults-centreon) [inline cloud/azure/database/redis/mode/cachehits.pm:111]:
```suggestion
Check Azure Redis cache usage statistics.
```

File: cloud/azure/database/redis/mode/cachelatency.pm [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    }
+}
+
+1;
+
+__END__
+
+=head1 MODE
+
+Check Azure Redis Cache latency.
```
Comment (thibaults-centreon) [inline cloud/azure/database/redis/mode/cachelatency.pm:103]:
```suggestion
Check Azure Redis cache latency statistics.
```

File: cloud/azure/database/redis/mode/cachethroughput.pm [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    }
+}
+
+1;
+
+__END__
+
+=head1 MODE
+
+Check Azure Redis Cache read/write throughput.
```
Comment (thibaults-centreon) [inline cloud/azure/database/redis/mode/cachethroughput.pm:111]:
```suggestion
Check Azure Redis cache throughput statistics.
```

File: cloud/azure/database/redis/mode/connectedclients.pm [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    }
+}
+
+1;
+
+__END__
+
+=head1 MODE
+
+Check Azure Redis Cache connected clients.
```
Comment (thibaults-centreon) [inline cloud/azure/database/redis/mode/connectedclients.pm:103]:
```suggestion
Check Azure Redis clients statistics.
```

File: cloud/azure/database/redis/mode/cpu.pm [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    }
+}
+
+1;
+
+__END__
+
+=head1 MODE
+
+Check Azure Redis Cache CPU usage.
```
Comment (thibaults-centreon) [inline cloud/azure/database/redis/mode/cpu.pm:103]:
```suggestion
Check Azure Redis CPU usage statistics.
```

File: cloud/azure/database/redis/mode/errors.pm [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    }
+}
+
+1;
+
+__END__
+
+=head1 MODE
+
+Check Azure Redis Cache errors.
```
Comment (thibaults-centreon) [inline cloud/azure/database/redis/mode/errors.pm:103]:
```suggestion
Check Azure Redis errors statistics.
```

File: cloud/azure/database/redis/mode/memory.pm [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+    }
+}
+
+1;
+
+__END__
+
+=head1 MODE
+
+Check Azure Redis Cache memory usage.
```
Comment (thibaults-centreon) [inline cloud/azure/database/redis/mode/memory.pm:103]:
```suggestion
Check Azure Redis memory statistics.
```

File: cloud/azure/database/redis/mode/serverload.pm [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+    }
+}
+
+1;
+
+__END__
+
+=head1 MODE
+
+Check Azure Redis Cache server load.
```
Comment (thibaults-centreon) [inline cloud/azure/database/redis/mode/serverload.pm:103]:
```suggestion
Check Azure Redis server load statistics.
```

**Review by thibaults-centreon — CHANGES_REQUESTED** [review]

**Comment by matoy** [comment]
i agreed all.

**Comment by thibaults-centreon** [comment]
Hello, changes related to my last comment have ben commited. I also added the "Health" mode that has been successfully tested. We're now good to go! Thanks for your contribution !

**Review by thibaults-centreon — APPROVED** [review]

**Pull request merged**